V1.1.0 for OmnAIView

AKMaily opened the pull request:
This is the first setup for OmnAIView v1.1.0 . This version is in non-production ready. 

It includes: 
        - a build prozess for windows 
        - typescript angular project 
        - typescript electron project 
        - MiniOmni v0.2.0 
        - the libusb.dll for MiniOmni 
        - modules to start an angular frontend
        - modules to build an electron app 

The first version has the following functionalitys: 
        - the backend : MiniOmni is started when the app is opened 
        - if the backend starts correctly it will be presented in a serverbox in the frontend 
        - if one or more OmnAIScopes are connected to the local device, those will be shown in the serverbox 
        - if selected and the button "Play" Button is pressed, the data from the device is shown in a plot on the top of the UI 
        - if selected and the button "Ausgewählte Geräte verbinden" is clicked, all selected devices are startedthe data from 
        all selected devices from all connected servers is shown in a plot on the top of the UI 
        - if selected and the button "Alle Geräte verbinden" is clicked all devices from all servers are started and their data is 
        shown in the plot 
      - if the button "Server hinzufügen" is clicked a popup menu is shown that has an input field for the port:id of another server. 
      This server can be added as a serverbox on the bottom of the UI 
      - If the software process is stopped also the backend process stops 

![image](https://github.com/user-attachments/assets/597b3763-2758-42e7-85b1-83f0861516a9)
![image](https://github.com/user-attachments/assets/73409761-a0d5-42f0-bf1a-6cefca6f29d8)

The goal of this PR is to review the code. If there are bugs in the software they will be set as an extra issue. 
If there is unclean code it should be changed before the PR is merged.

sigut001 committed:
refactor: reorganized interfaces and types in models.ts

- also little clean up

sigut001 committed:
fix(server): prevent parallel requests in device fetching

- Added `exhaustMap` to `timer` subscription to prevent race conditions
- Ensured new requests wait for the previous one to complete before executing

sigut001 committed:
refactor: Add explanations as comments

sigut001 committed:
feat(server): add configurable sampling rate for data downsampling

- Introduced `samplingRate` signal with a default value of 2000
- Updated `limitedData` computation to use `samplingRate` for dynamic downsampling
- Renamed `result` to `downsampledData` for better readability
- Added `setSamplingRate` method to allow runtime adjustments

sigut001 committed:
refactor: changed signal to readOnly because referenz never changes

sigut001 committed:
refactor: improve selectedDevices signal readability

- Updated `source` in `linkedSignal` to use a function for consistency
- Renamed `currentSelection` to `previousSelection` for consistency and clarity

sigut001 committed:
refactor: simplify numSelectedDevices computation

- Removed unnecessary variable assignment for `devices`
- Simplified `filter` condition to improve readability

sigut001 committed:
refactor: add possible todo with TODO-comment

sigut001 committed:
refactor: Add early return

sigut001 committed:
refactor: use constant for ignored message count

- Replaced hardcoded value `2` with `countOfFirstIgnoredMessages`

sigut001 committed:
refactor: mark properties as readonly

- Added `readonly` modifier to `httpClient`, `#servers`, and `servers`

sigut001 committed:
refactor: add missing }

sigut001 committed:
chore: translate comments in english

sigut001 committed:
refactor: rename functions and reorder codebase

- rename getDevicesURL() and getWSURL() to create-...
- rename getUUIDs() to fetchDevices(), and toggleDevice() to toggleSelectStateOfDevice(),
  to be more precise
- reorder variables of component to make it more readable
- remove todo-comment
- translate german comment
- streamline code to make it easer to read

sigut001 committed:
refactor: remove unused code, reorder and add readonly

sigut001 committed:
refactor: outsource tailwindstyles for ui-card

- add a seperate css file to make the code more readable
- remove german comment
- change space-y-4 to gap-4 for be more consitent in styling

sigut001 committed:
fix(ui-card): adjust width and alignment

- Set width to 100%
- Changed justify-content from center to start

sigut001 committed:
refactor: reorder serverlist and adjust styling

sigut001 committed:
feat(ui): improve server and device list layout

- Refactored container structure for better spacing and alignment
- Updated server list to a grid layout for improved responsiveness
- Applied consistent padding and gap adjustments
- Enhanced device list styling with clearer checkbox alignment
- Improved button styles for better user experience

sigut001 committed:
feat(data-service): add computed properties and device connection management

- Add computed properties to check server reachability and connection status
- Implement methods to enable, disable, toggle, connect, and disconnect devices across servers

sigut001 committed:
feat: add device selection helpers and improve connection logic

- Add `isDeviceSelected` computed property to check device selection status
- Implement `selectDevice` and `unselectDevice` methods for explicit device selection
- Modify connection logic to send only selected device UUIDs instead of all devices

sigut001 committed:
feat(device-selection): add methods to toggle device connections across servers

- Implement `toogleAllDevicesOfAllServers` to toggle all devices from all servers
- Implement `toogleCurrentDevicesFromAllServers` to connect currently chosen devices

sigut001 committed:
feat(ui): improve device and server management interface

- Add button group for managing server and device connections
- Implement dynamic button states based on server and device selection
- Enhance accessibility with better text contrast and styles
- Replace `selectedUUIDs().has(device.UUID)` with `serverObj.isDeviceSelected()(device.UUID)`

sigut001 committed:
fix: add missing return value

sigut001 committed:
chore: add padding

sigut001 committed:
refactor: change condition to outgray and disable Button

sigut001 committed:
refactor: change condition to outgray and disable Button

sigut001 committed:
Merge branch 'correctOrderOfServers' of https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron into correctOrderOfServers

sigut001 committed:
feat: add computed to get Status if any device is Available

sigut001 committed:
fix: Change condition about disabling "Alle GerÃ¤te verbinden"-button

- now its avaible if one Server is reachable and any device of them
is avaiable

sigut001 committed:
refactor: rename computed to improve semantic

- rename enableAllDevicesOfServer to selectAllDevicesOfServer
- rename disableAllDevicesOfServer to undselectAllDevicesOfServer

sigut001 committed:
fix: Add else statement for update the devicelist correctly

- in case that devices are disconnected from the the clients machine
  the devicelist in the ui also updates and removes the device.

sigut001 committed:
feat: refine computed properties for accurate state representation

- rename computed-variable to be more precise
- Implement `allDevicesSelected` to return true only when all devices are selected
- Add `allSelectedDevicesConnected` to indicate if all selected devices are connected
- Implement `allDevicesConnected` to return true only when all devices are selected and
the servers connected

sigut001 committed:
refactor(ui): centralize button state logic for better maintainability

- Introduce computed state variables for button disable conditions
- Consolidate `ngClass` bindings into predefined class variables
- Improve readability and maintainability of button state logic

sigut001 committed:
refactor(data-service): rename methods for clarity and consistency

- Rename `toogleAllDevicesOfAllServers` to `toggleAllServerDevicesBasedOnConnectionState`
- Rename `connectCurrentChoosenDevicesOfAllServers` to `connectSelectedDevicesOfAllServers`
- Rename `disconnectAllDevicesOfAllServers` to `disconnectAndUnselectAllDevicesOfAllServers`
- Remove redundant `disconnectCurrentChoosenDevicesOfAllServers` method

sigut001 committed:
chore: bump version to 1.0.5

sigut001 committed:
fix: changed version to 2.1.0

sigut001 committed:
fix: reflect renamed functionNames in codebase

sigut001 committed:
feat: add workflow

sigut001 committed:
refactor: remove outcomment functioncall to start backend

sigut001 committed:
refactor: remove seperate css classes and add tailwind instead

sigut001 committed:
fix: remove earlyreturn

- in this function the early-return is not necessary and also could
lead to bugs if the isConnected-var is not set to false

sigut001 committed:
refactor: add error handling for missing servers

- Throw an error when try to use a non-existent server
- Use this as early return to improve streamline of code

sigut001 committed:
refactor: change map to for each and comment

- change map to foreach to be precise in the usecase
- add comment wich describes that these functions have sideeffects

tabeatheunicorn committed:
Revert "feat: add workflow"

This reverts commit 2756451d215066c5f9c3fbcc7c586fad90af3ebd.

tabeatheunicorn committed:
add explanation to sr ratio

tabeatheunicorn committed:
fix: continue polling server after unreachable

When an error occurs in an rxjs Stream, it leads to the ending of the stream. By replacing it with an empty value,
the stream of fetching continues

tabeatheunicorn committed:
refactor: move UI in different components

As the UI components become more cluttered, it makes sense to divide them up into multiple different UI elements
which serve different taks. The dataservice is global so the referencing can  be easier done

tabeatheunicorn committed:
justify servers control buttons

Align text and icons so they look evenly spaced

AKMaily committed:
Update of electron 7.6.1 to 7.7.0

Changes:

- Updated electron 7.6.1 to 7.7.0
- deleted the plugin-auto-unpack-natives because using
	plugin-auto-unpack-natives with plugin fuses is not supported in
	the newest version . Both plugins try to start the app which is
	handled as an error in v7.7.0. As the App does not need native
	functions at the moment, this plugin doesnt seem to be used.

AKMaily committed:
Refactoring: Electron to TS Fix: Start BE local

Changes:
1. Setting up Electron from javascript to typescript for more type
   security
2. The local BE is now started in the installed app when starting the
   app
   -> Fix: For starting the BE from the installed app the wrong filepath
   was used, there was no seperation between the dev and production
   mode. Now there is a seperation between these two modes as well as
   between the used paths to the BE.

AKMaily committed:
Update the version number to 1.1.0

AKMaily committed:
Cleanup comments

tabeatheunicorn commented on angular-frontend/src/app/app.component.css:
we should delete empty files

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/485bb8c46ee99bc5a223bfe059bfb17f8724a91d

tabeatheunicorn commented on angular-frontend/src/app/app.component.html:
```diff
@@ -1,13 +1,13 @@
-  <aside class="h-full w-64 bg-omni-schwarz-700 text-white flex flex-col p-4  gap-8 rounded-r-lg">
+  <!-- Sidemenu -->
+  <aside
+    class="h-full w-64 bg-omni-schwarz-700 text-white flex flex-col p-4 gap-8 rounded-r-lg">
```
https://tailwindcss.com/docs/dark-mode
We should ensure that light mode and dark mode are separated

AKMaily replied:
I am not familiar with this part of the code and how to do that. Is it easy to implement?

tabeatheunicorn replied:
Well, tailwind allows you to use modifiers like dark:bg-dark-800, so when dark is true, it applies the selected class. This is more a general remark as I know that you prefer dark mode and the design might not take light/dark themes into account.

tabeatheunicorn replied:
So it is not hard to implement, but it is more time consuming as all styles need to be set twice

tabeatheunicorn commented on angular-frontend/src/app/app.component.html:
```diff
@@ -1,13 +1,13 @@
+  <!-- Main content -->
+  <main class="flex-1 flex flex-col p-4 gap-4">
+    <!-- controlbar top  -->
     <!-- <div
```
lets delete unused code

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/6199b705d93a71b2f76c72306766daaee122db2e

tabeatheunicorn commented on angular-frontend/src/app/app.component.ts:
https://stackoverflow.com/questions/39131675/is-it-good-practice-to-reference-services-in-html-templates-in-angular-2

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/data.models.ts:
```diff
@@ -0,0 +1,32 @@
+
+export interface DeviceInformation {
+  UUID: string;
+  color: { r: number; g: number; b: number };
```
this color model is used at least two times so it makes sense to create a datatype for it (especially as the type is not number but can be narrowed down further)

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/2c8cbd177925be2516b3ba027df673f92dd5dfe7

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/data.models.ts:
```diff
@@ -0,0 +1,32 @@
+  }[];
+}
+
+export interface omnAIScopeData {
```
we should document this dataformat in the code itself to ensure everybody knows how to use it

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/8a7f70a43c9e83fdf9941315647a7a14260e33fb

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/data.service.ts:
```diff
@@ -1,19 +1,23 @@
-  servers = this.#servers.asReadonly();
-
-  httpClient = inject(HttpClient);
+  readonly httpClient = inject(HttpClient);
```
this one should be private

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/data.service.ts:
```diff
@@ -39,7 +44,83 @@ export class DataService {
+   */
+  noServerReachable = linkedSignal(
+    () => {
+      const servers = Object.values(this.servers());
```
this line is written in many statements so we might want to add a signal for that

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/data.service.ts:
```diff
@@ -1,19 +1,23 @@
+import { DataFormat } from './data.models';
 
 @Injectable({ providedIn: 'root' })
 export class DataService {
```
The service is very long. We should try to shorten it and document when to use which function as well as paying attention to what is public and private

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/omnAIServer.ts:
```diff
@@ -1,68 +1,85 @@
   return `http://${serverURL}/UUID`;
 }
 
 export class ServerDescription {
```
This class feels a little unorganized, we might want to refactor given that this class is our interface with the backend and understanding it is an integral part of the app

tabeatheunicorn commented on angular-frontend/src/app/device-handling/device-selection/servers-control.component.ts:
```diff
@@ -0,0 +1,99 @@
+    });
+  }
+
+  disconnectAll() {
+    this.dataService.disconnectAndUnselectAllDevicesOfAllServers();
+  }
+
+  connectAll() {
+    this.dataService.connectAllDevicesOfAllServers();
+  }
+
+  connectSelected() {
+    this.dataService.connectSelectedDevicesOfAllServers();
+  }
```
These functions can be aliases

tabeatheunicorn commented on angular-frontend/src/app/device-handling/device-selection/servers-control.component.ts:
```diff
@@ -0,0 +1,99 @@
+  readonly dataService = inject(DataService);
+  readonly dialog = inject(MatDialog);
+
+  addNewDeviceServer() {
+    const dialogRef = this.dialog.open(AddDeviceServerComponent, {
+      minWidth: '30vw',
+    });
+
+    dialogRef.afterClosed().subscribe(result => {
+      if (result) {
+        this.dataService.addServer(result);
+      }
+    });
+  }
```
we might want to typecheck here

tabeatheunicorn commented on angular-frontend/src/app/device-handling/device-selection/servers-control.component.ts:
```diff
@@ -0,0 +1,99 @@
+
+        <button
+          class="rounded-md flex gap-2 items-center px-4 font-bold shadow-lg  text-slate-200 bg-green-500 disabled:bg-gray-500"
+          [disabled]="
+            dataService.noServerReachable() ||
+            dataService.noDeviceSelected() ||
+            dataService.allSelectedDevicesConnected() ||
+            dataService.allDevicesConnected()
+          "
```
These statements should be calculated in html so that it is easier to understand how the logic is derived and the html is not as cluttered

tabeatheunicorn commented on angular-frontend/src/app/device-handling/device-selection/servers-control.component.ts:
```diff
@@ -0,0 +1,99 @@
+  selector: 'app-servers-control',
+  standalone: true,
+  imports: [CommonModule, MatButtonModule],
+  template: `
```
We should move this to its own hmtl file as the template is getting longer

tabeatheunicorn commented on angular-frontend/src/app/device-handling/device-selection/servers-overview.component.ts:
```diff
@@ -0,0 +1,88 @@
+          class="bg-omni-schwarz-700 shadow-md rounded-lg p-4 relative min-h-[25vh]">
+          @if (!serverObj.serverIsReachable()) {
+            <div
+              class="absolute inset-0 w-full h-full bg-gray-600/60 flex items-center justify-center rounded-lg">
+              <span class="material-symbols-outlined text-white text-4xl mr-2"
+                >cloud_off</span
+              >
+              <span class="text-white text-lg">Server unerreichbar</span>
+            </div>
+          }
+
+          <div class="flex items-center justify-between px-4 py-s">
+            <h3 class="text-lg font-semibold text-slate-200">
+              {{ serverURL }}
+            </h3>
+            <button
+              [ngClass]="{
+                'text-green-500':
+                  !serverObj.isConnected() &&
+                  serverObj.numSelectedDevices() > 0,
+                'text-red-500': serverObj.isConnected(),
+                'text-gray-500':
+                  !serverObj.isConnected() &&
+                  serverObj.numSelectedDevices() === 0,
+              }"
+              class="px-4 py-2 rounded-md flex items-center"
+              (click)="
+                serverObj.isConnected()
+                  ? serverObj.disconnect()
+                  : serverObj.connect()
+              "
+              [disabled]="
+                !serverObj.isConnected() && serverObj.numSelectedDevices() === 0
+              ">
+              <span class="material-symbols-outlined text-4xl">
+                {{ serverObj.isConnected() ? 'stop_circle' : 'play_circle' }}
+              </span>
+            </button>
+          </div>
+
```
we should make this a server card as it is the view on one specific server and the overview should be able to select which view on the model is chosen

tabeatheunicorn commented on angular-frontend/src/index.html:
```diff
@@ -2,7 +2,7 @@
   <head>
     <meta charset="utf-8" />
-    <title>omnAIview</title>
+    <title>OmnAIview</title>
```
Please check with @M3hm3tB to ensure that the title is reflected in the electron app as well. He has already implemented that for other projects

tabeatheunicorn commented on electronTS/OmnAIView/package.json:
```diff
@@ -0,0 +1,48 @@
+  },
+  "keywords": [],
+  "author": {
+    "name": "AKMaily",
+    "email": "anna.theimann@outlook.de"
```
This should be updated to whoever signs it at the end

tabeatheunicorn commented on electronTS/OmnAIView/res/angular/3rdpartylicenses.txt:
artefacts should never be committed

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/08d9996803a48300547078e70480a461010cbb3d

https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/a3e6ee151f32a25f18bcde36d8a609bb2c99e7c7

tabeatheunicorn commented on electronTS/OmnAIView/src/main.ts:
```diff
@@ -0,0 +1,93 @@
+import * as path from "path";
+import * as fs from "fs-extra";
+
+// Prüfe, ob die App über Squirrel Startup gestartet wurde (Windows-Installer-Handling)
```
comments in english only

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/8597eb5349154313f7f47e4ea699db18aa335529

tabeatheunicorn commented on electronTS/OmnAIView/src/main.ts:
```diff
@@ -0,0 +1,93 @@
+// Prüfe, ob die App über Squirrel Startup gestartet wurde (Windows-Installer-Handling)
+if (require("electron-squirrel-startup")) app.quit();
+
+let backendProcess: ChildProcess | null = null; // Backend-Prozess-Referenz
```
do you need a global handle for that?

AKMaily replied:
What would be a better way to integrate the backendProcess in the startBackend and stopBackend functions ?

tabeatheunicorn commented on electronTS/OmnAIView/src/main.ts:
```diff
@@ -0,0 +1,93 @@
+
+let backendProcess: ChildProcess | null = null; // Backend-Prozess-Referenz
+
+// Backend starten
```
these comments are obsolet

tabeatheunicorn commented on electronTS/OmnAIView/src/main.ts:
```diff
@@ -0,0 +1,93 @@
+}
+
+// Electron-Hauptfenster erstellen
+function createWindow(): void {
+  const mainWindow: BrowserWindow = new BrowserWindow({
+    width: 800,
+    height: 600,
+    webPreferences: {
+      preload: path.join(__dirname, "preload.js"),
+      contextIsolation: true,
+      nodeIntegration: false,
+    },
+    autoHideMenuBar: true,
+    icon: path.join(__dirname, "res", "Icon.ico"),
+  });
+
+  const indexPath: string = path.join(__dirname, "..", "res", "angular", "browser", "index.html");
+  mainWindow.loadFile(indexPath).catch(err => console.error("Fehler beim Laden der HTML-Datei:", err));
+}
```
we might want to define that in a renderer.ts file as this code can be separate from the main app setttings

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/a3a164575708bddc6d592e7cb29ebac6b813e80e

https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/24d23fa9f72b03fc9df870c448b4c8168aaa7ba9

tabeatheunicorn commented on electronTS/OmnAIView/src/main.ts:
```diff
@@ -0,0 +1,93 @@
+});
+
+ipcMain.handle("run-omnai-command", (event, commandArgs: string[]) => {
+  const exePath: string = app.isPackaged
+  ? path.join(process.resourcesPath, "MiniOmni.exe")
+  : path.join(__dirname, "res", "omnai", "MiniOmni.exe");
+
```
This should be an extra function as the logic will change

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/ca76fc186c8404037e39085621ec15b9979a6c9e

tabeatheunicorn commented on electronTS/OmnAIView/src/preload.ts:
```diff
@@ -0,0 +1,38 @@
+    onClosed: (callback: (message: string) => void) => void;
+  }
+  
+  // Sicheren Zugriff auf APIs gewähren
```
Please translate comments

AKMaily replied:
https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/8597eb5349154313f7f47e4ea699db18aa335529

tabeatheunicorn commented on electronTS/OmnAIView/src/preload.ts:
```diff
@@ -0,0 +1,38 @@
+    sendCommand: (commandArgs: string[]) => ipcRenderer.invoke("run-omnai-command", commandArgs),
+  
+    // Startet den Backend-Prozess
+    startBackend: () => ipcRenderer.invoke("start-backend"),
```
this looks weird - why did you write this?

tabeatheunicorn commented on electronTS/OmnAIView/src/preload.ts:
```diff
@@ -0,0 +1,38 @@
+// https://www.electronjs.org/docs/latest/tutorial/process-model#preload-scripts
+import { contextBridge, ipcRenderer, IpcRendererEvent } from "electron";
+
+interface OmnAIAPI {
```
I am not sure why you would write this interface? is it needed to get rid of the invoke commands in main.ts?

AKMaily replied:
Turns out the preload.ts was not used in the first place . 
Fixed in : https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron/pull/4/commits/485bb8c46ee99bc5a223bfe059bfb17f8724a91d

AKMaily committed:
Deletion of not used code

AKMaily committed:
Updated comments to english. OmnAIView instead of Omnaiview

AKMaily committed:
Update .gitignore

Updated .gitignore to ignore res, out and dist folder in electron

AKMaily committed:
Ignore res folder

AKMaily committed:
Deletion of not used files

AKMaily committed:
.gitignore only for res/angular

AKMaily committed:
Update: getBackendPath-Function

Function to get the path to the backend. If the path changes it only has
to be changed in the function instead of at a few places in the code .

AKMaily committed:
RGB Datatyp for the colors

As the color is used a few times in the code it should be declared as
a datatype. If at some point the dataformat for the color is changed, it
can be updated easier. Also it is more readable where this type comes
from.

AKMaily committed:
Documentation of the omnAIScopeData interface

As the omnaiscopedata format could change and its interface is used in several
functions in the app, it should be documented how the interface works
and what component it has.
Therefore the interface omnAIScopeDate is documented in the code .

AKMaily committed:
Enhancement: Renderer.ts

Define the createWindow function from the main.ts to seperate the
rendering settings from the main app settings.

AKMaily committed:
Enhancement: Smaller main.ts

Deleted the ipcMain functions as they were not used.
Moved the backend function for the app start into backend.ts, which
makes the code more readable and structured.
